### PR TITLE
feat: Add transfer account authority ix

### PR DESCRIPTION
**Pull request by exdx**
This PR adds the ability to transfer account authority to another account via a feature flag. This provides the ability to migrate accounts on the platform.

### CODE REVIEW

**Comment by exdx** [comment]
Not ready for review.

**Review by exdx — COMMENTED** [review]
Pretty small set of changes if this is all we need. 

I don't see a way to set flags from the CLI -- I'm wondering if we wanted to add that? Could be a separate PR

File: programs/marginfi/src/state/marginfi_account.rs [hunk 1]
```diff
@@ -77,6 +78,24 @@ impl MarginfiAccount {
     pub fn get_flag(&self, flag: u64) -> bool {
         self.account_flags & flag != 0
     }
+
+    pub fn set_new_account_authority(
+        &mut self,
+        new_authority: Pubkey,
+    ) -> std::result::Result<(), MarginfiError> {
```
[the remark this answers is not in the document]
Comment (jkbpvsc) [reply programs/marginfi/src/state/marginfi_account.rs:85]: Signature is good, you can use a type alias we already have for this Result in `create::prelude`.
I would add to the name of the fn to `_checked` so its clear there are conditions to the change.

File: programs/marginfi/src/state/marginfi_account.rs [hunk 2]
```diff
@@ -77,6 +78,24 @@ impl MarginfiAccount {
...
+        // check if new account authority flag is set
+        if !self.get_flag(TRANSFER_AUTHORITY_ALLOWED_FLAG) || self.get_flag(DISABLED_FLAG) {
+            return Err(MarginfiError::IllegalAccountAuthorityTransfer);
```
[the remark this answers is not in the document]
Comment (jkbpvsc) [reply programs/marginfi/src/state/marginfi_account.rs:88]: Yes that is the assumption we are working with. 
Last thing your code is missing is the actual instruction itself, the account layout (and security declaration) and including the instruction in the `#[program]` in the `lib.rs`

**Comment by jkbpvsc** [comment]
If you can also add a few integration tests that cover theses cases:
- Non owner tries to change the account authority -> fail
- Non whitelisted (no flag set) owner tries to change the authority -> fail
- Whitelisted owner tries to change the authority -> success

File: programs/marginfi/src/instructions/marginfi_account/transfer_authority.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+pub struct MarginfiAccountSetAccountAuthority<'info> {
+    #[account(mut)]
+    pub marginfi_account: AccountLoader<'info, MarginfiAccount>,
+
+    #[account(
+        address = marginfi_account.load()?.authority,
+    )]
+    pub signer: Signer<'info>,
+
+    pub new_authority: AccountLoader<'info, MarginfiAccount>,
```
Comment (jkbpvsc) [inline programs/marginfi/src/instructions/marginfi_account/transfer_authority.rs:46]: The `new_authority` is not a `MarginfiAccount` account, its just an address which we will check as the signer moving forward.

Meaning we should use 
```
 pub new_authority: AccountInfo<'info>,
```

File: programs/marginfi/src/instructions/marginfi_account/transfer_authority.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    state::marginfi_account::MarginfiAccount,
+};
+use anchor_lang::prelude::*;
+
+pub fn set_account_transfer_authority(
+    ctx: Context<MarginfiAccountSetAccountAuthority>,
+) -> MarginfiResult {
+    // Gather accounts
+    let mut marginfi_account = ctx.accounts.marginfi_account.load_mut()?;
+    let new_account_authority = ctx.accounts.new_authority.load()?;
```
Comment (jkbpvsc) [inline programs/marginfi/src/instructions/marginfi_account/transfer_authority.rs:13]: No need to load this as if its an account loader, bc we are only interested in the address of the account info

File: programs/marginfi/tests/marginfi_account.rs [hunk 5]
```diff
@@ -2410,3 +2411,195 @@ async fn lending_account_close_balance() -> anyhow::Result<()> {
...
+    let transfer_account_authority_ix = Instruction {
+        program_id: marginfi::id(),
+        accounts: marginfi::instructions::MarginfiAccountSetAccountAuthority {
+            marginfi_account: todo!(),
+            signer: todo!(),
+            new_authority: todo!(),
+            fee_payer: todo!(),
+        }
+        .to_account_metas(None),
+        data: vec![],
```
Comment (jkbpvsc) [inline programs/marginfi/tests/marginfi_account.rs:2475]: Can you add a helper function to the MarginfiAccountFixture `try_transfer_account_authority` so we are not duplicating the logic across tests

**Review by jkbpvsc — CHANGES_REQUESTED** [review]

**Comment by jkbpvsc** [comment]
Also seems tests have failed

File: programs/marginfi/tests/marginfi_account.rs [hunk 6]
```diff
@@ -2410,3 +2411,195 @@ async fn lending_account_close_balance() -> anyhow::Result<()> {
...
+    assert_eq!(marginfi_account.authority, test_f.payer());
+    // transfer flag not set on the new account
+    assert!(!marginfi_account.get_flag(TRANSFER_AUTHORITY_ALLOWED_FLAG));
+    let new_account_authority_pk: [u8; 32] = [0; 32];
+
+    // // attempt authority transfer
+    // TODO: fix types
+    let transfer_account_authority_ix = Instruction {
+        program_id: marginfi::id(),
+        accounts: marginfi::instructions::MarginfiAccountSetAccountAuthority {
```
Comment (jkbpvsc) [inline programs/marginfi/tests/marginfi_account.rs:2468]: Shoule be marginfi::accounts::MarginfiAccountSetAccountAuthority

File: programs/marginfi/tests/marginfi_account.rs (hunk 5 above)
Comment (jkbpvsc) [inline programs/marginfi/tests/marginfi_account.rs:2475]: should be marginfi::instruction::MarginfiAccountSetAccountAuthority

**Comment by exdx** [comment]
Implemented the new functionality and it looks really good now. Running into some CI issues though:
```
error[E0080]: evaluation of constant value failed
  --> programs/marginfi/src/state/marginfi_account.rs:26:1
   |
26 | assert_struct_size!(MarginfiAccount, 2304);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ attempt to compute `0_usize - 1_usize`, which would overflow
   |
   = note: this error originates in the macro `const_assert` which comes from the expansion of the macro `assert_struct_size` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I get this error when building on an M1 mac but they went away on an x86 box. I don't think any struct sizes have changed? Trying to see if I changed something

**Comment by exdx** [comment]
The issue is that the `MarginfiAccountSetAccountAuthority` accounts type is too large for the stack. It's an anchor issue. Looking into how to slim down the size of the struct by boxing accounts

**Review by jkbpvsc — CHANGES_REQUESTED** [review]

File: test-utils/src/marginfi_account.rs [hunk 7]
```diff
@@ -625,4 +625,39 @@ impl MarginfiAccountFixture {
     pub fn get_size() -> usize {
         mem::size_of::<MarginfiAccount>() + 8
     }
+
+    /// Use the client to send the transfer ix authority transaction
+    pub async fn try_transfer_account_authority(
+        &self,
```
Comment (jkbpvsc) [inline test-utils/src/marginfi_account.rs:631]: `new_authority` should probably be a fn argument here

Comment (exdx) [reply test-utils/src/marginfi_account.rs:631]: I agree. I seem to break the test every time I add this though lol. 😅

**Comment by exdx** [comment]
Seems like the CI problems here are across open PRs and not related to this PR's functionality.

**Comment by exdx** [comment]
The fuzz test is broken across CI currently -- not related to the changes made in this PR.

**Review by jkbpvsc — APPROVED** [review]

**Pull request merged**